protocol/triple/internal/dubbo3_server/api/greet_service.go

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api
+
+import (
+	"context"
+	"dubbo.apache.org/dubbo-go/v3/protocol/triple/internal/proto"
+	greet "dubbo.apache.org/dubbo-go/v3/protocol/triple/internal/proto/dubbo3_gen"
+	"fmt"
+	"github.com/dubbogo/gost/log/logger"
+	"github.com/pkg/errors"
+	"io"
+	"strings"
+)
+
+type GreetDubbo3Server struct {
+	greet.UnimplementedGreetServiceServer
+}
+
+func (srv *GreetDubbo3Server) Greet(ctx context.Context, req *proto.GreetRequest) (*proto.GreetResponse, error) {
+	return &proto.GreetResponse{Greeting: req.Name}, nil
+}
+
+func (srv *GreetDubbo3Server) GreetStream(stream greet.GreetService_GreetStreamServer) error {
+	for {
+		req, err := stream.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			logger.Errorf("Stream Recv unexpected err: %s", err)
+			return fmt.Errorf("dubbo3 recv error: %s", err)
+		}
+		if err := stream.Send(&proto.GreetStreamResponse{Greeting: req.Name}); err != nil {
+			logger.Errorf("Stream Send unexpected err: %s", err)
+			return fmt.Errorf("dubbo3 send error: %s", err)
+		}
+	}
+	return nil
+}
+
+func (srv *GreetDubbo3Server) GreetClientStream(stream greet.GreetService_GreetClientStreamServer) error {
+	var reqs []string
+	for {
+		req, err := stream.Recv()
```

Comment:
If we introduce ended and write code this way, ended should be set true or false when receiving unexpected error from client-side or server-side? If we set true, then users have to examine the type of err and take related action. If we set false, users also have to examine err to determine whether they should go on.

Maybe we could provide a function named with triple.IsEnd(err error)?
Then code would be like this:
req, err := stream.Recv()
if err != nil {
    if triple.IsEnd(err) {
        // do something
    }
    // process unexpected error
}
// Send your request here

Reply:
And this interface is generated by old triple, we can not change it...

Reply:
I didn't type all code above. 

```go
req, err, ended := stream.Recv()
if err != nil {
    // Handle error
}
if ended {
    // Do something
} else {
    // Send your request here
}
```

Wrapping a method, named `isEnded`, to the user looks good to me, since "this interface is generated by old triple".

Reply:
In general, gRPC stream and classic network programming would return EOF to represent the ending. So I am not sure that users would treat **triple.IsEnded** as a better way.

Based on the code above, maybe we should still return **io.EOF** as error when we design **Stream** interface for our new Triple? Then ended flag seems redundant.

Reply:
My key point is that the user don't need to know what is `io.EOF`, and don't need to know what's happening under the hood of gRPC. You know `io.EOF` since you are familiar with gRPC. However, we can't expect all users to understand these things . We should provide intuitive things to the user.

IMO, the ideal returned values are:

- err: errors expect for `io.EOF` indicating gRPC or user business went wrong;
- ended: equals to `io.EOF`, but it is a bool value;
- req: request.

But it might be not implemented due to the old Triple implementation. Wdyt?

Reply:
For users who are not familiar with gRPC or some classic network programing at all, **ended** is intuitive to them because ending of stream is a very common case and should not be treated as error (IMO).

But the key point there is that should we return **io.EOF** when we supply users with **ended** flag? If we set err with **io.EOF** and **ended** with true, it would seems tedious for experienced users. If we set err with nil and ended with true, we still need to explain this change to them.

// general use
req, err := stream.Recv()
if err != nil {
    if triple.IsEnded(err) {
        // do something
    }
    // process unexpected error
}
// Send your request here

// experienced users
req, err := stream.Recv()
if err != nil {
    if errors.Is(err, io.EOF) {
        // do something
    }
    // process unexpected error
}
// Send your request here

// triple.IsEnded
func IsEnded(err error) bool {
    return errors.Is(err, io.EOF)
}
This code would not increase the cost of understanding for experienced users. Beginners just need to learn triple.IsEnded.

Reply:
Fair enough. Could you add this function into this patch?

Reply:
Ack.